### PR TITLE
Fix issue 4520 due to storage model mismatch

### PR DESCRIPTION
**Pull request by sklam**
as titled.

### CODE REVIEW

**Comment by esc** [comment]
@sklam this is dark. Can you give us a `--verbose` on why this fixes the issue?

**Comment by sklam** [comment]
@esc, yes, i'll try to explain this.

**Background**

Numba has this concept called *datamodel* (which I decided to call in storage model in the PR title because it is probably more descriptive).  The *datamodels* describe how a type is represented in different situations:

- *value* representation is for in-registers data or temporary in-stack data that does not need to be addressable.
- *data* representation is for heap data that needs to be addressable.
- *argument* and *return* representations need to be ABI compatible for the corresponding function boundary.

Most types have the same *value* and *data* representation but a bool is stored as a single bit in the *value* representation and a single byte in the *data* representation. The reason being that LLVM uses a `i1` (1-bit integer) as the predicate for the branch instruction `br`, but a `i1` is not addressable. The smaller addressable unit is a byte `i8`.

Related code: https://github.com/numba/numba/blob/master/numba/datamodel/models.py

**The bug**

This patch fixed the incref/decref wrapper function for the containers which is taking a pointer to the container elements using the *data* representation, but the underlying incref/decref operations are expecting the *value* representation.  Elements in containers must be stored in the *data* representation because it needs to be addressable. To convert the *data* representation to *value* representation, the patch used the `datamodel.load_from_data_pointer()` instead of the `builder.load()`.

File: numba/tests/test_dictobject.py [hunk 1]
```diff
@@ -1229,6 +1229,22 @@ def foo(d):
         # Value is correctly updated
         self.assertPreciseEqual(d[1], np.arange(10, dtype=np.int64) + 100)
 
+    def test_storage_model_mismatch(self):
```
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py]: These tests need a python 2 guard as they use the string type.

Comment (sklam) [reply numba/tests/test_dictobject.py]: done

**Pull request merged**